From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Newwind <support@newwindserver.com>
Date: Sat, 27 Jul 2024 11:09:37 +0200
Subject: [PATCH] Prevent spawn placement loading chunks

isSpawnPositionOk() can causes sync chunk loads under specific circumstances, one known example is when
a zombie calls reinforcements and checks isSpawnPositionOk in 50 positions within a 40 block radius
This patch ensures the block state is loaded before checking if mobs can spawn on it.


diff --git a/src/main/java/net/minecraft/world/entity/SpawnPlacementTypes.java b/src/main/java/net/minecraft/world/entity/SpawnPlacementTypes.java
index f718edd11..bdd9c0dd5 100644
--- a/src/main/java/net/minecraft/world/entity/SpawnPlacementTypes.java
+++ b/src/main/java/net/minecraft/world/entity/SpawnPlacementTypes.java
@@ -27,7 +27,8 @@ public interface SpawnPlacementTypes {
             if (entityType != null && world.getWorldBorder().isWithinBounds(pos)) {
                 BlockPos blockPos = pos.above();
                 BlockPos blockPos2 = pos.below();
-                BlockState blockState = world.getBlockState(blockPos2);
+                BlockState blockState = world.getBlockStateIfLoaded(blockPos2); // Paper - Prevent spawn placements from loading chunks
+                if (blockState == null) return false; // Paper - Prevent spawn placements from loading chunks
                 return blockState.isValidSpawn(world, blockPos2, entityType)
                     && this.isValidEmptySpawnBlock(world, pos, entityType)
                     && this.isValidEmptySpawnBlock(world, blockPos, entityType);
